### PR TITLE
GEODE-1617: Regions can be created with a variety of characters that are unsupported

### PR DESCRIPTION
**Pull request by kjduling**
Updated the pull request for this ticket after several unit tests failed.

### CODE REVIEW

**Comment by kirklund** [comment]
I'm running this change through precheckin.

**Comment by jinmeiliao** [comment]
I think it's probably safe to get the storage team involved to review this PR.

**Comment by kirklund** [comment]
There are still a lot of failures including most of the geode-lucene tests (apparently they use # character in all of their region names). See http://precheckin.gemfire.pivotalci.info:8080/pipelines/gemfire-precheckin-klund-precheckin-geode-1617-pr-201-003/jobs/DeveloperTest/builds/1

File: geode-core/src/main/java/com/gemstone/gemfire/cache/RegionFactory.java [hunk 1]
```diff
@@ -823,6 +825,38 @@ public void setPublisher(boolean v) {
...
+   *           if timed out getting distributed lock for Scope.GLOBAL
+   * @throws CacheClosedException
+   *           if the cache is closed
+   * @throws IllegalStateException
+   *           if the supplied RegionAttributes are incompatible with this region
+   *           in another cache in the distributed system (see
+   *           {@link AttributesFactory} for compatibility rules)
+   */
+  @SuppressWarnings("unchecked")
+  public Region<K,V> create(String name, RegionAttributes<K, V> regionAttributes, InternalRegionArguments internalRegionArguments)
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/com/gemstone/gemfire/cache/RegionFactory.java]: RegionFactory is an external API. So you can't add this "create" method to it because it takes "InternalRegionArguments" which is not part of the external API.

File: geode-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java [hunk 2]
```diff
@@ -17,10 +17,96 @@
 
 package com.gemstone.gemfire.internal.cache;
 
-import com.gemstone.gemfire.*;
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java]: If you want to clean up these imports you should not do it as part of this fix. File a new ticket for it. I could be wrong about this but I think you should change your IDE to not automatically reorg the imports. Kirk might have some good advice on this.

File: geode-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java [hunk 3]
```diff
@@ -7769,17 +7879,55 @@ void cleanUpOnIncompleteOp(EntryEventImpl event, RegionEntry re) {
...
+//        if (metaRegion.getName().equals(name)) {
+//          return;
+//        }
+//      }
+//      if (metaRegion instanceof SerialGatewaySenderQueueMetaRegion) {
+//        return;
+//      }
+//    }
+
+    if (name.startsWith("__")
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java]: Why was the special case needed? Shouldn't this region have set isInternalRegion?

File: geode-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java [hunk 4]
```diff
@@ -7769,17 +7879,55 @@ void cleanUpOnIncompleteOp(EntryEventImpl event, RegionEntry re) {
...
+        && !name.equals(ClientHealthMonitoringRegion.ADMIN_REGION_NAME)) {
+      throw new IllegalArgumentException("Region names may not begin with a double-underscore: " + name);
+    }
+
+    // Ensure the region only contains valid characters
+    Pattern pattern = Pattern.compile("[aA-zZ0-9-_.]+");
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java]: Would it be worth putting this Pattern in a static final so that we don't compile it on every name check?

File: geode-core/src/main/java/com/gemstone/gemfire/internal/cache/wan/serial/SerialGatewaySenderQueue.java [hunk 5]
```diff
@@ -979,15 +1009,14 @@ private void initializeRegion(AbstractGatewaySender sender,
...
-                  .setImageTarget(null)
-                  .setIsUsedForSerialGatewaySenderQueue(true)
-                  .setSerialGatewaySender(sender));
-          
+          this.region = gemCache.createVMRegion(this.regionName, ra, new InternalRegionArguments().setInternalMetaRegion(meta)
+                                                                                                  .setDestroyLockFlag(true)
+                                                                                                  .setSnapshotInputStream(null)
+                                                                                                  .setImageTarget(null)
+                                                                                                  .setIsUsedForSerialGatewaySenderQueue(true)
+                                                                                                  .setInternalRegion(true)
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/com/gemstone/gemfire/internal/cache/wan/serial/SerialGatewaySenderQueue.java]: I would be nice if this diff just showed one line (the addition of ".setInternalRegion(true)". Are you not using the checked in geode code style? I actually like your new indentation better but it seems like we should all use a consistent formatter.

File: geode-core/src/main/java/org/apache/geode/redis/GeodeRedisServer.java [hunk 6]
```diff
@@ -424,14 +429,26 @@ private void initializeRedis() {
...
-        redisMetaData = rfMeta.create(REDIS_META_DATA_REGION);
+      try {
+        if ((stringsRegion = this.cache.getRegion(STRING_REGION)) == null)
+          stringsRegion = rfString.create(GeodeRedisServer.STRING_REGION, ra, ira);
+        if ((hLLRegion = this.cache.getRegion(HLL_REGION)) == null)
+          hLLRegion = rfHLL.create(HLL_REGION, ra, ira);
+        if ((redisMetaData = this.cache.getRegion(REDIS_META_DATA_REGION)) == null)
+          redisMetaData = rfMeta.create(REDIS_META_DATA_REGION, ra, ira);
+      }
+      catch (IOException | ClassNotFoundException e) {
```
Comment (dschneider-pivotal) [inline geode-core/src/main/java/org/apache/geode/redis/GeodeRedisServer.java]: The addition of this try/catch does not look related to region names. Seems like it should have its own ticket.

File: geode-core/src/test/java/com/gemstone/gemfire/cache/query/QueryJUnitTest.java [hunk 7]
```diff
@@ -117,6 +118,7 @@ public void test002UnicodeInQuery() {
   }
 
   @Test
+  @Ignore // GEODE-1617 - Regions can be created with a variety of characters that are unsupported
```
Comment (dschneider-pivotal) [inline geode-core/src/test/java/com/gemstone/gemfire/cache/query/QueryJUnitTest.java]: Shouldn't these tests be removed instead of ignored?

**Comment by seankarson** [comment]
The example in the docs for how to create a dynamic region on the server no longer works in M3 due to this fix.

http://geode.docs.pivotal.io/docs/developing/region_options/dynamic_region_creation.html

Removing an underscore from "__regionAttributesMetadata" does the trick.

I'm happy to make a PR against the docs to fix this, but it doesn't seem like there is an M3 branch

**Comment by kjduling** [comment]
Good catch.

@davebarnes97 @karensmolermiller Can either of you follow up on this?  Looks like we've a page out of date.

**Comment by joeymcallister** [comment]
@seankarson @kjduling The document has been updated. Thanks for the heads up!
